044-kraft-liveness-readiness.md

```diff
@@ -0,0 +1,198 @@
+* Update the KafkaAgent to check for a broker state of >= 3 && != 127.
+This will be used for the broker only and combined readiness checks later and will improve the existing ZooKeeper based broker check.
+
+### Phase 2 - Strimzi adds support for broker only and controller only modes
+* Controller only pod liveness and readiness checks are fully implemented to match this proposal.
+* Broker only pods are marked as both alive and ready if listening on port 9091.
+* Combined mode pods are marked as both alive and ready when there is a Java process running (no change from phase 1).
+
+### Phase 3 - KafkaRoller is updated to take the existence of controller pods and status of controller quorum into account when deciding whether to roll pods
+* Broker only readiness probe fully implemented to match this proposal.
+* Combined mode readiness probe fully implemented to match this proposal.
+
+## Affected/not affected projects
+
+This change only affects the Cluster Operator and Kafka brokers.
+
+## Compatibility
+
+As part of this change the readiness probe for a Kafka broker in a ZooKeeper based cluster will be updated to 
+continue being marked as ready when the BrokerState metric > 3 && != 127.
+This means it will tolerate the PENDING_CONTROLLED_SHUTDOWN and SHUTTING_DOWN states.
+This will not cause any compatibility problems because currently the KafkaAgent stops running once the metric reaches 
+the RUNNING state.
+The change will make the code between KRaft and ZooKeeper mode simpler and protect readiness if in future the agent is 
+updated to continue running once the broker is "ready".
+
+## Rejected alternatives
+
+### Use the current-state metric for controller readiness
+The [current-state](https://kafka.apache.org/documentation/#kraft_quorum_monitoring) metric indicates the current state of the quorum.
+Previous versions of this proposal discussed whether we could check if this metric had a state of either follower or leader and use that to infer that the controller quorum had been formed and the controller should be marked as ready.
+This approach had a couple of problems:
+* The individual state of the controller doesn't always imply the full quorum state. 
+For example, this particular controller might be offline, but the rest of the quorum is still healthy.
+Also, during a leadership election the metric might show a state of `candidate` or `voted`.
+* In general readiness checks in Kubernetes should be based on individual nodes, not the cluster as a whole.
+* The current-state metric is not accessible as a Yammer metric, so cannot be accessed by the KafkaAgent in the same way it currently reads the BrokerState metric.
+Although it could perhaps be read by the KafkaAgent if it called the MBean server within the JVM directly or read it from `$LOG_DIR/__cluster_metadata-0/quorum-state`.
+* The broker nodes will not move out of the STARTING state until the quorum leader election has happened.
```

Comment:
Just a formatting change

```suggestion
### Use the current-state metric for controller readiness

The [current-state](https://kafka.apache.org/documentation/#kraft_quorum_monitoring) metric indicates the current state of the quorum.
Previous versions of this proposal discussed whether we could check if this metric had a state of either follower or leader and use that to infer that the controller quorum had been formed and the controller should be marked as ready.
This approach had a couple of problems:
* The individual state of the controller doesn't always imply the full quorum state. 
  For example, this particular controller might be offline, but the rest of the quorum is still healthy.
  Also, during a leadership election the metric might show a state of `candidate` or `voted`.
* In general readiness checks in Kubernetes should be based on individual nodes, not the cluster as a whole.
* The current-state metric is not accessible as a Yammer metric, so cannot be accessed by the KafkaAgent in the same way it currently reads the BrokerState metric.
  Although it could perhaps be read by the KafkaAgent if it called the MBean server within the JVM directly or read it from `$LOG_DIR/__cluster_metadata-0/quorum-state`.
* The broker nodes will not move out of the STARTING state until the quorum leader election has happened.
```